sdk/servicebus/service-bus/src/shared/openLink.ts

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+import * as log from "../log";
+import { Receiver as RheaReceiver, AwaitableSender } from "rhea-promise";
+import { translate, defaultLock } from "@azure/core-amqp";
+import { AbortSignalLike } from "@azure/core-http";
+import { AbortError } from "@azure/abort-controller";
+
+export type RheaLink = AwaitableSender | RheaReceiver;
+
+/**
+ * This is transitional code - as more code gets moved out of MessageReceiver/MessageSender
+ * we can stop passing in some of these functions.
+ *
+ * For instance
+ * - negotiateClaim + ensureTokenRenewal can be lifted out of LinkEntity
+ * - openLock can be moved into here if we allow openLink to have some manageable state
+ *  - same for isConnecting, which might need to be readable external from here but not writable.
+ *
+ * @internal
+ * @ignore
+ */
+export interface OpenArgs<RheaLinkT extends RheaLink> {
+  create(): Promise<RheaLinkT>;
+  isOpen(): boolean;
+
+  // logging related stuff.
+  logger: typeof log.sender;
+  logPrefix: string;
+
+  // TODO: I'd like to let openLink have some state (object?) and then these would be internal details
+  // we could still let external callers check if isConnecting is set, but not change it.
+  openLock: string;
+
+  isConnecting(): boolean;
+  setIsConnecting(value: boolean): boolean;
+
+  getCloseInitiated(): boolean;
+
+  // these are more candidates to move out from LinkEntity. They don't really depend on anything special
+  // in LinkEntity and are just clutter.
+  negotiateClaim(): Promise<void>;
+  ensureTokenRenewal(): Promise<void>;
+
+  abortSignal?: AbortSignalLike;
+}
+
+/**
+ * @internal
+ * @ignore
+ */
+export function openLink<T extends RheaLink>(args: OpenArgs<T>): Promise<T | undefined> {
+  const checkAborted = (): void => {
+    if (args.abortSignal?.aborted) {
+      throw new AbortError("open() has been cancelled by user.");
+    }
+  };
+
+  checkAborted();
+
+  if (args.isOpen() || args.isConnecting() || args.getCloseInitiated()) {
+    return Promise.resolve(undefined);
+  }
+
+  // TODO: set isConnecting up here, cleanup on finally?
+
+  args.logger(
+    `${args.logPrefix} is not open or connecting. Acquiring lock ${args.openLock} for initializing the session, link and possibly the connection.`
+  );
+
+  // TODO: do we _need_ this lock at the moment? If we're not trying to get
+  // open and close to be mutually exclusive?
+  return defaultLock.acquire(args.openLock, async () => {
+    checkAborted();
```

Comment:
There are a lot of calls to `checkAborted()` being done in here after acquiring the lock. I understand the need to use it so that we can exit as soon as possible instead of going through the different async operations, but we don't need it to throw the AbortError itself. 

We can use `checkAndRegisterWithAbortSignal()` before calling `openLink()`. This way, in here, we only need to add early exits before every async operation. And then when these async operations can support abort signal in the future, these checks can go away.